include/deal.II/fe/fe_p1nc.h

```diff
@@ -0,0 +1,311 @@
+ *
+ *  @verbatim
+ *  2---------|---------3
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  -                   -
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  0---------|---------1
+ *  @endverbatim
+ *
+ * For each vertex $v_j$ of given quad, there are two edges of which $v_j$ is one of end points.
+ * Consider a linear function such that 0.5 value at two midpoints of such edges,
+ * and 0.0 at two midpoints of other edges.
+ * Note that the set of these values satisfies the dice rule which is described above.
+ * We denote such a function associated with vertex $v_j$ by $\phi_j$.
+ *
+ * The following figures represent $\phi_j$, $j=0,\cdots,3$ with its values at midpoints.
+ * Canonical (local) basis functions are given by any three shape functions among
+ * the following four linear functions.
+ *
+ * <ul>
+ * <li> shape function $\phi_0$:
+ *  @verbatim
+ *  +--------0.0--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.5                 0.0
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.5--------+
+ *  @endverbatim
+ *
+ * <li> shape function $\phi_1$:
+ *  @verbatim
+ *  +--------0.0--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.0                 0.5
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.5--------+
+ *  @endverbatim
+ *
+ * <li> shape function $\phi_2$:
+ *  @verbatim
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.5                 0.0
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+ *
+ * <li> shape function $\phi_3$:
+ *  @verbatim
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.0                 0.5
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+ *
+ * </ul>
+ *
+ * Note that above shape functions are constructed on each cell, not on the reference cell only.
+ * @p get_linear_shape computes the coefficients for shape functions when @p fill_fe_values
+ * is called on each cell.
+ *
+ * The basis function associated with a node in global scope is defined by the composition of
+ * (local) basis functions associated with the node on each element.
+ * And DOF associated with the node represents the coefficient of the basis function associated with the node.
+ * In this frame, all DOFs are truly independent.
```

Comment:
It is a 'weird' property of P1NC.

When we consider only a single cell, 4 'local' shape functions are linearly dependent because the dimension of the linear polynomial space is just 3.
If we consider given global mesh, we define 'global' basis functions for the global function space. Each of 'global' basis functions is associated with a node and defined by a cell-wise composition of 'local' shape functions on cells containing that node as a vertex.

In short, the linear independency of those 'global' basis functions depends on a type of given boundary condition.
They are linearly independent when the homogeneous Dirichlet BC is given. In this case, the 'global' basis functions associated with 'interior nodes' are linearly independent. And the number of 'global' DOFs is same as the number of interior nodes.
But in case of Neumann BC - I think that a situation of a mesh with only one cell is included in this case because all 'boundary nodes' are considered - it is more complicated. The 'global' basis functions associated with 'all nodes'  are actually not linearly independent, they has 1 redundancy relation. Thus the number of DOFs is same as the number of all nodes -1.